### PR TITLE
Changing backend framework to Flask

### LINKED ISSUE #68: Migrate to Flask
Responder was great to get off the ground, but now that this a real project we should treat it as such and use something more people can contribute to.

- [x] be able to hit hello world endpoints
- [x] get slack webhook working
- [x] make a migration
	- [x] alembic will have to point to different `db` object for migrations
	- [x] look into flask-migrate (this could be a project for later)
- [x] connect flask endpoints to tasks we kick off in task queue
- [x] implement github oauth flow

### PR DESCRIPTION
**Pull request by alysivji**
Closes #68 

This is a pretty hefty PR. It's a lot easier to rip the bandaid off versus trying to migrate an endpoint at a time for a project that does 2 things.

### Reason for Change

Busy-Beaver has been glitching; Responder eats Exceptions; it's been difficult to get a peek into the system. Plus with new contributors coming on board, it makes sense to use popular frameworks that have well-known behaviors.

### High-Level Overview

Migrate the backend from responder, sqlalchemy-wrapper, and alembic to Flask, flask-sqlalchemy, and flask-migrate.

We are now using `gunicorn` versus `uvicorn` in both dev and prod docker-compose environments.

#### Flask

- Fairly textbook use of Flask, straight out of docs
- Making use of [application factories](http://flask.pocoo.org/docs/1.0/patterns/appfactories/), [blueprints](http://flask.pocoo.org/docs/1.0/blueprints/), [error handlers](http://flask.pocoo.org/docs/1.0/patterns/errorpages/#error-handlers), and [other patterns outlined in the docs](http://flask.pocoo.org/docs/1.0/patterns/)
- class-based views that inherit from `MethodView`

Some of my customizations to make life easy:
- each request has an `_internal` dictionary that we can store request-specific data inside
- decorator to allow token-based authentication with roles (ported from #62)
	- I don't like how we hook this up to views, but it's good enough for now. There are many other things to get done. If this becomes a problem, we can use a flask plugin to manage API user accounts 

#### Database

- removed code in migration which required Twitter integration
- removed Key-Value table created by `simplekv` and replaced with a relational database that is wrapped using a `KeyValueStoreAdapter`
	- we have to keep track of single values for various workspaces (i.e. what was the last tweet sent, what was the last video updated). Currently, we have this implemented as a table for a single tenant. as we expand to more workspaces, we can add a column to specify the workspace if it has enabled the integration

#### Other improvements

- pytest fixtures helpers. the `session` fixture improves interaction with test database; database transactions are local to each test. transaction starts when each test begins and is rolled back once each test is complete
	- check out `conftest.py` and how tests are structured, I'm a fan. Would love to talk about other patterns / ideas that improve the experience
- the `request_client` adapter is now more robust
	- initailize object with user-supplied headers
	- perform post requests (!!!!)

### Todo

- asynchronous task queue with redis and rq (high priority)
- flask-admin or another admin interface (medium priority)
- marshaling endpoints with marshmallow (low priority)

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -41,24 +41,27 @@ down:
...
+	docker-compose exec app flask db upgrade
 
-migrate-down: ## Rollback migrations using alembic
-	docker-compose exec app alembic --config=./migrations/alembic.ini downgrade -1
+migrate-down: ## Rollback migrations
+	docker-compose exec app flask db downgrade
 
 test:
-	docker-compose exec app pytest
+	docker-compose exec app pytest -k "$(k)"
```
Comment (alysivji) [inline Makefile]: Note: make more general in next PR so we can just pass in command line options

File: busy_beaver/blueprints/integration/github.py [hunk 2]
```diff
@@ -1,70 +1,65 @@
...
 logger = logging.getLogger(__name__)
+client = RequestsClient()
 
-CLIENT_ID = os.getenv("GITHUB_APP_CLIENT_ID")
-CLIENT_SECRET = os.getenv("GITHUB_APP_CLIENT_SECRET")
-GITHUB_REDIRECT_URI = "https://busybeaver.sivji.com/github-integration"
 
-
-class GitHubIdentityVerificationCallbackResource:
+class GitHubIdentityVerificationCallbackResource(MethodView):
```
Comment (alysivji) [inline busy_beaver/blueprints/integration/github.py]: This is currently testing manually. Automation is on my todo list.

File: busy_beaver/blueprints/tasks/__init__.py [hunk 3]
```diff
@@ -0,0 +1,18 @@
+from flask import blueprints
+
+from .github_summary import PublishGitHubSummaryResource
+from .retweeter import TwitterPollingResource
+from busy_beaver.decorators import authentication_required
+
+tasks_bp = blueprints.Blueprint("tasks", __name__)
+admin_role_required = authentication_required(roles=["admin"])
```
Comment (alysivji) [inline busy_beaver/blueprints/tasks/__init__.py]: As we are using class-based views, this is how we have to decorate views to add token-based authentication. I am not a fan, but the workflow is completely tested and works exactly as expected.

Being pragmatic and getting things working in the way they were before. We can always add a flask plugin to manage our API User accounts and roles.

File: busy_beaver/config.py [hunk 4]
```diff
@@ -1,6 +1,10 @@
 import os
 
 IN_PRODUCTION = os.getenv("IN_PRODUCTION", False)
+PROD_BASE_URI = "https://busybeaver.sivji.com"
+DEV_BASE_URI = os.getenv("NGROK_BASE_URI", None)
```
Comment (alysivji) [inline busy_beaver/config.py]: note to self: add `NGROK_BASE_URI` to contributor guidelines (future PR)

File: busy_beaver/decorators.py [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+                logger.error("[Busy-Beaver] Unauthorized access of endpoint")
+                data = {"message": "Not authorized to access endpoint, contact admin"}
+                return make_response(401, error=data)
+
+            request._internal["user"] = api_user
+            return func(*args, **kwargs)
+
+        # TODO Not sure why we need this line... can funtools fix it?
+        # https://stackoverflow.com/questions/17256602/assertionerror-view-function-mapping-is-overwriting-an-existing-endpoint-functi
+        _token_auth.__name__ = func.__name__
```
Comment (alysivji) [inline busy_beaver/decorators.py]: `¯\_(ツ)_/¯`

File: busy_beaver/exceptions.py [hunk 6]
```diff
@@ -4,3 +4,13 @@ class BusyBeaverException(Exception):
 
 class UnexpectedStatusCode(BusyBeaverException):
     pass
+
+
+class NotFoundError(BusyBeaverException):
```
Comment (alysivji) [inline busy_beaver/exceptions.py]: not used, but wanted to get this pattern out there

File: busy_beaver/toolbox.py [hunk 7]
```diff
@@ -1,7 +1,27 @@
...
 import pytz
 
 
 def utc_now_minus(period: timedelta):
     return pytz.utc.localize(datetime.utcnow()) - period
+
+
+def make_response(
+    status_code: int = 200, *, headers: dict = {}, json: dict = {}, error: dict = {}
+):
```
Comment (alysivji) [inline busy_beaver/toolbox.py]: I do this for most of my projects. I'm not sure what good practice is, but it works well for me. Thoughts?

File: tests/blueprints/integration/slack_test.py [hunk 8]
```diff
@@ -0,0 +1,173 @@
...
+@pytest.fixture
+def patcher(monkeypatch):
+    """Helper to patch in the correct spot"""
+
+    def _patcher(namespace, replacement_object):
+        namespace_to_patch = f"{MODULE_TO_TEST}.{namespace}"
+        monkeypatch.setattr(namespace_to_patch, replacement_object)
+        return replacement_object
+
+    yield _patcher
```
Comment (alysivji) [inline tests/blueprints/integration/slack_test.py]: This is something new I'm trying out. Still figuring it out. Lemme know if you like using it.

Basic idea is that it patches the exact namespace you want, which is always one of my biggest gripes with `mock.patch()`

File: tests/conftest.py [hunk 9]
```diff
@@ -1,6 +1,68 @@
+"""Project wide pytest plugins / fixtures
+
+References:
+    - http://alexmic.net/flask-sqlalchemy-pytest/
```
Comment (alysivji) [inline tests/conftest.py]: S/O to the blog post I borrowed the idea from

File: busy_beaver/decorators.py (hunk 5 above)
Comment (thornycrackers) [reply busy_beaver/decorators.py]: @alysivji  I think functools wraps will help with this

File: busy_beaver/toolbox.py (hunk 7 above)
Comment (thornycrackers) [reply busy_beaver/toolbox.py]: I like it, makes it easy.

File: tests/blueprints/integration/slack_test.py (hunk 8 above)
Comment (thornycrackers) [reply tests/blueprints/integration/slack_test.py]: I'm not sure I understand how this one works, I'll have to get some hands on with it to see how it works

File: busy_beaver/toolbox.py (hunk 7 above)
Comment (alysivji) [reply busy_beaver/toolbox.py]: Note: should set default values to `None`. If the value is still `None` inside the function, we can create an empty dictionary. This way can produce unexpected values as we are creating an object when this function is first read by the interpreter.

Currently, we are not seeing issues as the dictionary is not mutated, but it's better Python to change it. Do it in a separate PR so we can show that not everything is perfect and that we all screw up.

**Pull request merged**